proposals/extensions.md

```diff
@@ -313,8 +313,115 @@ However, beyond implementing the language semantics it must satisfy certain requ
 
 These requirements need more refinement as implementation progresses, and may need to be compromised in corner cases in order to allow for a reasonable implementation approach.
 
+### Metadata for declarations
+
+Each extension container is emitted as a nested public static class with a marker method.  
+
+Its name is unspeakable and determined based on the lexical order in the program. 
+The name is not guaranteed to remain stable across re-compilation. 
+Below we use `<Extension>ee_` followed by an index. For example: `<Extension>ee_2`.  
+
+Method/property/indexer declarations in an extension container in source are represented as skeleton declarations in metadata.  
+The signatures of the original methods are maintained, but their bodies are replaced with `throw null`.  
+Those should not be referenced in IL.  
+
+Note: This is similar to ref assemblies. The reason for using `throw null` bodies (as opposed to no bodies) 
+is so that IL verification could run and pass (thus validating the completeness of the metadata).
+
+The extension marker method encodes the receiver parameter.  
+- It is public and static, and is called `<Extension>$`.  
+- If the receiver parameter doesn't specify a name, then the parameter name is unspeakable.  
+
+Note: This allows roundtripping of extension container symbols through metadata (full and reference assemblies).  
+
+The method bodies for method/property/indexer declarations in an extension container in source are emitted 
+as static implementation methods in the top-level static class.  
+- The name of an implementation method prepends `<Extension>` to the name of the original method.  
+  For example: `set_Property` => `<Extension>set_Property`.
+- An implementation method has type parameters derived from the extension container prepended to the type parameters of the original method.  
+- An implementation method has the same accessibility as the original method.  
+- The implementation method corresponding to a static method has the same parameters and return type as the original method. 
+- The implementation method corresponding to an instance method adds a prepended parameter to the signature of the original method. 
```

Comment:
How do we avoid a possible name and signature collision with static extension method after this parameter is added to the signature?

Reply:
The name was munged to avoid collisions. If we choose to keep speakable names instead, we'll likely have to stick a modopt or a rule to avoid collisions.

Reply:
Sorry, I'd misunderstood. The problem was with `static void M(IEnumerable<T>)` and `void M()`. From offline discussion, this is now solved by using a different prefix for static methods (`<StaticExtension>`).